src/libraries/Microsoft.PowerFx.Core/Public/Values/RecordValue.cs

```diff
@@ -15,48 +15,109 @@ namespace Microsoft.PowerFx.Core.Public.Values
+        {
+        }
 
         internal RecordValue(IRContext irContext)
             : base(irContext)
         {
             Contract.Assert(IRContext.ResultType is RecordType);
         }
 
-        // If we have a derived Value, we can get a derived type. 
+        /// <summary>
+        /// The RecordType of this value. 
+        /// </summary>
         public new RecordType Type => (RecordType)base.Type;
 
         public static RecordValue Empty()
         {
             var type = new RecordType();
             return new InMemoryRecordValue(IRContext.NotInSource(type), new List<NamedValue>());
         }
-         
+
         /// <summary>
-        /// Get a field on this record. 
-        /// See <see cref="ObjectMarshaller"/> to override GetField behavior. 
+        /// Get a field on this record.         
         /// </summary>
-        /// <param name="name"></param>
-        /// <returns></returns>
-        public FormulaValue GetField(string name)
-        {            
-            var fieldType = Type.MaybeGetFieldType(name) ?? FormulaType.Blank;
+        /// <param name="fieldName">Name of field on this record.</param>
+        /// <returns>Field value or blank if missing. </returns>
+        public FormulaValue GetField(string fieldName)
+        {
+            if (fieldName == null)
+            {
+                throw new ArgumentNullException(nameof(fieldName));
+            }
 
-            return GetField(IRContext.NotInSource(fieldType), name);
+            var fieldType = Type.MaybeGetFieldType(fieldName) ?? FormulaType.Blank;
+
+            return GetField(fieldType, fieldName);
         }
 
-        internal virtual FormulaValue GetField(IRContext irContext, string name)
+        // Internal, for already verified values.
+        internal FormulaValue GetField(FormulaType fieldType, string fieldName)
+        {
+            if (TryGetField(fieldType, fieldName, out var result))
+            {
+                if (result == null)
+                {
+                    throw new InvalidOperationException($"{GetType().Name}.TryGetField({fieldName}) returned true but null.");
+                }
+
+                Contract.Assert(result.Type.Equals(fieldType) || result is ErrorValue || result.Type is BlankType);
+
+                return result;
+            }
+
+            // This should be a compiler-error. 
+            // Fx semantics are to return blank on missing fields. 
+            return FormulaValue.NewBlank(fieldType);
+        }
+
+        /// <summary>
+        /// Derived classes can override. 
+        /// </summary>
+        /// <param name="fieldType"></param>
+        /// <param name="fieldName"></param>
+        /// <returns></returns>
+        protected virtual bool TryGetField(FormulaType fieldType, string fieldName, out FormulaValue result)
         {
             // Derived class can have more optimized lookup.
             foreach (var field in Fields)
```

Comment:
>    foreach (var field in Fields) [](http://example.com/codeflow?start=8&length=33)

I have a problem - 
there's a circular reference between TryGetField --> Fields --> TryGetFields. 

TryGetFields should be abstract.

Reply:
fixed in latest.